### PR TITLE
Add random normal operation

### PR DESCRIPTION
**Pull request by doru1004**
Add random normal operation.

### CODE REVIEW

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 1]
```diff
@@ -163,6 +163,41 @@ static FlatSymbolRefAttr getOrInsertMemcpy(
   return SymbolRefAttr::get(context, "llvm.memcpy.p0i8.p0i8.i64");
 }
 
+static FlatSymbolRefAttr getOrInsertRandomNormal(
+    PatternRewriter &rewriter, ModuleOp module, mlir::Type inType) {
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:167]: nit: I do not thing the namespace is required (mlir::Type --> Type).

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 2]
```diff
@@ -163,6 +163,41 @@ static FlatSymbolRefAttr getOrInsertMemcpy(
...
+  auto *context = module.getContext();
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:168]: [Suggestion]: Using `auto` is convenient but does not convey the static type to readers. The LLVM code style rule encourage to not use `auto` when the type is obvious:
```
Don’t “almost always” use auto, but do use auto with initializers like cast<Foo>(...) or other places where the type is already obvious from the context"

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 3]
```diff
@@ -163,6 +163,41 @@ static FlatSymbolRefAttr getOrInsertMemcpy(
...
+  StringRef functionName = inType.isF64() ? "get_random_normal_value_f64"
+                                          : "get_random_normal_value_f32";
+  if (module.lookupSymbol<LLVM::LLVMFuncOp>(functionName.str()))
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:171]: nit: I have see the `llvm` name space only used in lower cases.

File: src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===----------------- Matmul.cpp - Lowering Matmul Op --------------------===//
+//
+// Copyright 2019 The IBM Research Authors.
```
Comment (etiotto) [inline src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp:7]: Do we need the copyright?  If we do fix 2019 -> 2021  (time flies when working on a fun project :-))

File: src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/Krnl/KrnlHelper.hpp"
+#include "src/Dialect/ONNX/IndexExpr.hpp"
+#include "src/Dialect/ONNX/MLIRDialectBuilder.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+
+#include <time.h>
```
Comment (etiotto) [inline src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp:21]: #include \<ctime\>

File: src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+using namespace mlir;
+
+#define DEBUG_TRACE 0
```
Comment (etiotto) [inline src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp:25]: I do not see this macro used.  Consider using LLVM_DEBUG infrastructure for traces.

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 7]
```diff
@@ -1363,6 +1398,46 @@ class KrnlVectorTypeCastOpLowering : public ConvertToLLVMPattern {
...
+// KRNL to LLVM: KrnlRandomNormalOpLowering
+//===----------------------------------------------------------------------===//
+
+class KrnlRandomNormalOpLowering : public ConversionPattern {
+public:
+  explicit KrnlRandomNormalOpLowering(MLIRContext *context)
+      : ConversionPattern(KrnlRandomNormalOp::getOperationName(), 1, context) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:1411]: override -> final ?

File: src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+struct ONNXRandomNormalOpLowering : public ConversionPattern {
+  ONNXRandomNormalOpLowering(MLIRContext *ctx)
+      : ConversionPattern(
+            mlir::ONNXRandomNormalOp::getOperationName(), 1, ctx) {}
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    Location loc = op->getLoc();
+    auto outputMemRefType = convertToMemRefType(*op->result_type_begin());
```
Comment (etiotto) [inline src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp:34]: `auto` everywhere ... consider using `auto` only where the type is (a) obvious or (b) unwieldy like when declaring a lambda function.

File: src/Runtime/OMRandomNormal.c [hunk 9]
```diff
@@ -0,0 +1,15 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------- OMRandomNormal.cpp - OMRandomNormal C Implementation ---------===//
+//
+// Copyright 2019-2020 The IBM Research Authors.
```
Comment (etiotto) [inline src/Runtime/OMRandomNormal.c:7]: 2019-2020  --> 2021

File: src/Runtime/OMRandomNormal.cpp [hunk 10]
```diff
@@ -0,0 +1,15 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------- OMRandomNormal.c - OMRandomNormal C++ Implementation ---------===//
+//
+// Copyright 2019-2020 The IBM Research Authors.
```
Comment (etiotto) [inline src/Runtime/OMRandomNormal.cpp:7]: 2019-2020 --> 2021

File: src/Runtime/OMRandomNormal.inc [hunk 11]
```diff
@@ -0,0 +1,62 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------ OMRandomNormal.inc - OMRandomNormal C/C++ Implementation ------===//
+//
+// Copyright 2019-2020 The IBM Research Authors.
```
Comment (etiotto) [inline src/Runtime/OMRandomNormal.inc:7]: 2019-2020 --> 2021

File: src/Runtime/OMRandomNormal.inc [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+// This file contains C/C++ implementation of the OMRandomNormal functions.
+//
+//===----------------------------------------------------------------------===//
+
+
+#ifdef __cplusplus
+
+#include <random>
+
+void get_random_normal_value_f64(double *result, int64_t size, double mean, double scale, double seed) {
```
Comment (etiotto) [inline src/Runtime/OMRandomNormal.inc:20]: Suspiciously long line: can you try to run clang-format.

File: src/Runtime/OMRandomNormal.inc [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    result[index] = normalRandom()*scale+mean;
+}
+
+void get_random_normal_value_f32(float *result, int64_t size, float mean, float scale, float seed) {
+  srand(seed);
+  for(int64_t index=0; index<size; ++index)
+    result[index] = normalRandom()*scale+mean;
+}
+
+#endif
```
Comment (etiotto) [inline src/Runtime/OMRandomNormal.inc:62]: add newline at end of file

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp (hunk 2 above)
Comment (AlexandreEichenberger) [reply src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp]: personally agree with minimizing auto.

File: src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp (hunk 4 above)
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/Math/RandomNormal.cpp]: Name is also RandomNormal.cpp

**Review by AlexandreEichenberger — COMMENTED** [review]
looks good once comments are addressed

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 14]
```diff
@@ -163,6 +163,41 @@ static FlatSymbolRefAttr getOrInsertMemcpy(
...
+    llvmOptionsTy = FloatType::getF64(context);
+    llvmOutputTy = LLVM::LLVMPointerType::get(llvmOptionsTy);
+  }
+  auto llvmI64Ty = IntegerType::get(context, 64);
+  auto llvmFnType = LLVM::LLVMFunctionType::get(llvmVoidTy,
+      ArrayRef<mlir::Type>({llvmOutputTy, llvmI64Ty, llvmOptionsTy,
+          llvmOptionsTy, llvmOptionsTy}),
+      false);
+
+  // Insert the memcpy function into the body of the parent module.
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:193]: The comment mentions memcpy but you are creating the function type for "get_random_normal_value_f[64|32]".

**Pull request merged**